### PR TITLE
Add MoLeRGenerator model

### PR DESCRIPTION
**Pull request by sarahnlewis**
Add the MoLeRGenerator model which samples molecules autoregressively using the MoLeR decoder. Refactor MoLeRVae so that I can do this without too much code duplication.

### CODE REVIEW

File: molecule_generation/models/moler_generator.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+                input_molecule_representations=molecule_representations,
+                graphs_requiring_node_choices=batch_features[
+                    "partial_graphs_requiring_node_choices"
+                ],
+                candidate_edges=batch_features["valid_edge_choices"],
+                candidate_edge_features=batch_features["edge_features"],
+                candidate_attachment_points=batch_features["valid_attachment_point_choices"],
+            ),
+            training=training,
+        )
```
[the remark this answers is not in the document]
Comment (sarahnlewis) [reply molecule_generation/models/moler_generator.py]: Done

File: molecule_generation/models/moler_base_model.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+        )
+
+        total_loss = (
+            +self._params["node_classification_loss_weight"]
+            * decoder_metrics.node_classification_loss
+            + self._params["first_node_classification_loss_weight"]
+            * decoder_metrics.first_node_classification_loss
+            + self._params["edge_selection_loss_weight"] * decoder_metrics.edge_loss
+            + self._params["edge_type_loss_weight"] * decoder_metrics.edge_type_loss
+        )
```
[the remark this answers is not in the document]
Comment (sarahnlewis) [reply molecule_generation/models/moler_base_model.py]: Fixed

File: molecule_generation/models/moler_base_model.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+    def decoder_prefix(cls) -> str:
+        return cls.__decoder_prefix
+
+    @classmethod
+    def get_default_hyperparameters(cls, mp_style: Optional[str] = None) -> Dict[str, Any]:
+        base_hypers = super().get_default_hyperparameters(mp_style)
+        base_hypers.update(
+            {
+                # Dimension of the context/conditioning vector supplied to the MoLeRDecoder.
+                # Actually, we set this vector to all-zeros, but leave it for ease of reuse of MoLeR VAE code
```
[the remark this answers is not in the document]
Comment (sarahnlewis) [reply molecule_generation/models/moler_base_model.py]: I rewrote this comment.

File: molecule_generation/layers/moler_decoder.py [hunk 4]
```diff
@@ -1674,9 +1687,10 @@ def generate_result_that_picks_carbon():
 
     def _decoder_pick_new_bond_types(
         self,
+        *,
+        sampling_mode: DecoderSamplingMode,
         decoder_states: List[MoLeRDecoderState],
         store_generation_traces: bool = False,
-        sampling_mode: DecoderSamplingMode = DecoderSamplingMode.GREEDY,
         num_samples: int = 1,
     ) -> Iterable[
```
[the remark this answers is not in the document]
Comment (sarahnlewis) [reply molecule_generation/layers/moler_decoder.py]: Fixed. This was just a failure to finish tidying up, sorry.

File: molecule_generation/cli/train.py [hunk 5]
```diff
@@ -105,13 +106,21 @@
...
     model_default_hypers={"num_train_steps_between_valid": 5000},
 )
 
+cli.register_task(
+    task_name="MoLeRGenerator",
+    dataset_class=JSONLMoLeRTraceDataset,
+    dataset_default_hypers=moler_dataset_params,
+    model_class=MoLeRGenerator,
+    model_default_hypers={"num_train_steps_between_valid": 5000},
+)
```
[the remark this answers is not in the document]
Comment (sarahnlewis) [reply molecule_generation/cli/train.py:122]: If I'm looking at the right runs for [MoLeRGenerator](https://ml.azure.com/experiments/id/23a4f4bf-25f6-4dcf-9759-39867ae90904/runs/2fe67777-f793-4859-bde1-084b5d0309bf?wsid=/subscriptions/681d5936-9572-4a8f-aa6f-8f9d42a8abc4/resourceGroups/genchem_ci/providers/Microsoft.MachineLearningServices/workspaces/genchem_ci&tid=72f988bf-86f1-41af-91ab-2d7cd011db47#) and [MoLeRVae](https://ml.azure.com/runs/d5bed37d-7456-4239-a9c1-a9119c801f85?wsid=/subscriptions/681d5936-9572-4a8f-aa6f-8f9d42a8abc4/resourceGroups/genchem_ci/providers/Microsoft.MachineLearningServices/workspaces/genchem_ci&tid=72f988bf-86f1-41af-91ab-2d7cd011db47#) then MoLeRVae and MoLeRGenerator do similarly on FCD and KL divergence, but MoLeRGenerator does badly on uniqueness and novelty.

File: molecule_generation/cli/train.py [hunk 6]
```diff
@@ -144,12 +153,12 @@ def log(msg) -> None:
             f"but got {loaded_model_dataset[0]}!"
         )
     dataset: JSONLAbstractTraceDataset = loaded_model_dataset[0]
-    if not isinstance(loaded_model_dataset[1], (CGVAE, MoLeRVae)):
+    if not isinstance(loaded_model_dataset[1], (CGVAE, MoLeRVae, MoLeRGenerator)):
         raise ValueError(
             f"This training script can only work with CGVAE/MoLeRVae, "
```
Comment (kmaziarz) [inline molecule_generation/cli/train.py:158]:
```suggestion
            f"This training script can only work with CGVAE/MoLeR, "
```

File: molecule_generation/models/moler_vae.py [hunk 7]
```diff
@@ -1,6 +1,7 @@
 """MoLeR Variational Autoencoder model."""
 from dataclasses import dataclass
-from typing import Any, Dict, Optional, Tuple, List, NamedTuple, Union, Iterable
+from typing import Any, Dict, Optional, Tuple, List, NamedTuple, Union
+from molecule_generation.models.moler_base_model import MoLeRBaseModel
```
Comment (kmaziarz) [inline molecule_generation/models/moler_vae.py:4]: nit: put this together with the other `from molecule_generation` imports

File: molecule_generation/test/models/test_moler_vae.py [hunk 8]
```diff
@@ -8,6 +8,7 @@
 from dpu_utils.utils import RichPath
 from tf2_gnn import DataFold
 
+from molecule_generation.models.moler_generator import MoLeRGenerator
 from molecule_generation.dataset.jsonl_moler_trace_dataset import JSONLMoLeRTraceDataset
 from molecule_generation.models.moler_vae import MoLeRVae
```
Comment (kmaziarz) [inline molecule_generation/test/models/test_moler_vae.py:13]: nit: sort imports
```suggestion
from molecule_generation.dataset.jsonl_moler_trace_dataset import JSONLMoLeRTraceDataset
from molecule_generation.models.moler_generator import MoLeRGenerator
from molecule_generation.models.moler_vae import MoLeRVae
```

File: molecule_generation/utils/model_utils.py [hunk 9]
```diff
@@ -4,6 +4,7 @@
 
 from tf2_gnn.cli_utils.model_utils import get_model, get_model_file_path, load_weights_verbosely
 
+from molecule_generation.models.moler_generator import MoLeRGenerator
 from molecule_generation.models.cgvae import CGVAE
 from molecule_generation.models.moler_vae import MoLeRVae
 from molecule_generation.dataset.in_memory_trace_dataset import InMemoryTraceDataset
```
Comment (kmaziarz) [inline molecule_generation/utils/model_utils.py:10]: nit: sort imports
```suggestion
from molecule_generation.dataset.in_memory_trace_dataset import InMemoryTraceDataset
from molecule_generation.models.cgvae import CGVAE
from molecule_generation.models.moler_generator import MoLeRGenerator
from molecule_generation.models.moler_vae import MoLeRVae
```

File: molecule_generation/utils/model_utils.py [hunk 10]
```diff
@@ -53,7 +54,7 @@ def load_vae_model_and_dataset(
     weight_file = get_model_file_path(trained_model_path, "hdf5")
     load_weights_verbosely(weight_file, model)
 
-    if not (isinstance(model, CGVAE) or isinstance(model, MoLeRVae)):
+    if not isinstance(model, (CGVAE, MoLeRVae, MoLeRGenerator)):
         raise ValueError(f"Model loaded from {trained_model_path} not a CGVAE or MoLeRVae model!")
```
Comment (kmaziarz) [inline molecule_generation/utils/model_utils.py:58]:
```suggestion
        raise ValueError(f"Model loaded from {trained_model_path} not a CGVAE or MoLeR model!")
```

File: molecule_generation/utils/moler_inference_server.py [hunk 11]
```diff
@@ -11,6 +11,8 @@
 from rdkit import Chem
 from more_itertools import chunked, ichunked
 
+from molecule_generation.models.moler_generator import MoLeRGenerator
+from molecule_generation.utils.moler_decoding_utils import DecoderSamplingMode
 from molecule_generation.dataset.in_memory_trace_dataset import InMemoryTraceDataset, DataFold
 from molecule_generation.models.moler_vae import MoLeRVae
 from molecule_generation.utils.moler_decoding_utils import MoLeRDecoderState
```
Comment (kmaziarz) [inline molecule_generation/utils/moler_inference_server.py:18]: nit: sort and combine imports
```suggestion
from molecule_generation.dataset.in_memory_trace_dataset import InMemoryTraceDataset, DataFold
from molecule_generation.models.moler_generator import MoLeRGenerator
from molecule_generation.models.moler_vae import MoLeRVae
from molecule_generation.utils.moler_decoding_utils import DecoderSamplingMode, MoLeRDecoderState
```

**Review by kmaziarz — CHANGES_REQUESTED** [review]
LGTM, thanks!

(I took a final pass, only noting a few small nitpicks. The ones related to import order/grouping aren't enforced in CI yet, but _mostly_ followed in _most_ files...)

File: molecule_generation/models/moler_vae.py (hunk 7 above)
Comment (sarahnlewis) [reply molecule_generation/models/moler_vae.py]: Done

**Review by kmaziarz — APPROVED** [review]

**Comment by kmaziarz** [comment]
Actually, one more thing: I guess we need to update `README.md` to replace the uses of `ModelWrapper` with `VaeWrapper`, as the former is now an abstract base class that cannot be instantiated?

**Comment by sarahnlewis** [comment]
> Actually, one more thing: I guess we need to update `README.md` to replace the uses of `ModelWrapper` with `VaeWrapper`, as the former is now an abstract base class that cannot be instantiated?

Done. I also added `VaeWrapper` and `GeneratorWrapper` to `__init__.py`.

**Pull request merged**